pallets/parachain-staking/src/lib.rs

```diff
@@ -1947,84 +2025,138 @@ pub mod pallet {
+		/// NOTE: It is assumed that the calling context checks whether the
 		/// collator candidate is currently active before calling this function.
 		///
 		/// # <weight>
-		/// Weight: O(D) where D is the number of top candidates.
-		/// - Reads: TopCandidates
-		/// - Writes: TopCandidates
+		/// Weight: O(1)
+		/// - Reads: TopCandidates, CandidatePool, TotalCollatorStake
+		/// - Writes: TopCandidates, TotalCollatorStake
 		/// # </weight>
-		fn update_top_candidates(candidate: T::AccountId, old_total: BalanceOf<T>, new_total: BalanceOf<T>) {
-			// check if candidate is in top_candidates (and update)
-			// or if candidate will ascend into top_candidates
-			let mut top_candidates = <TopCandidates<T>>::get();
+		fn update_top_candidates(
+			candidate: T::AccountId,
+			old_self: BalanceOf<T>,
+			old_delegators: BalanceOf<T>,
+			new_self: BalanceOf<T>,
+			new_delegators: BalanceOf<T>,
+		) -> u32 {
+			// check if candidate is in TopCandidates (and update)
+			// or if candidate will ascend into TopCandidates
+			let mut top_candidates = TopCandidates::<T>::get();
+			let num_top_candidates: u32 = top_candidates.len().saturated_into();
 			let old_stake = Stake {
 				owner: candidate.clone(),
-				amount: old_total,
+				amount: old_self.saturating_add(old_delegators),
 			};
+			let new_total = new_self.saturating_add(new_delegators);
+
+			// case 1: candidate has been a member of TopCandidates
 			if let Ok(i) = top_candidates.linear_search(&old_stake) {
 				top_candidates.mutate(|vec| {
 					if let Some(stake) = vec.get_mut(i) {
 						stake.amount = new_total;
 					}
 				});
 				TopCandidates::<T>::put(top_candidates);
+
+				// update total amount at stake only if candidate is collator
+				if i.saturated_into::<u32>() < MaxSelectedCandidates::<T>::get() {
+					TotalCollatorStake::<T>::mutate(|total| {
+						total.collators = total.collators.saturating_sub(old_self).saturating_add(new_self);
+						total.delegators = total
+							.delegators
+							.saturating_sub(old_delegators)
+							.saturating_add(new_delegators);
+					});
+				}
+			// case 2: candidate newly joins TopCandidates and either replaces
+			// another candidate in the list or updates their own stake
 			} else if let Ok(drop_out) = top_candidates.try_insert_replace(Stake {
 				owner: candidate.clone(),
 				amount: new_total,
 			}) {
-				if let Some(drop_out) = drop_out {
-					Self::deposit_event(Event::LeftTopCandidates(drop_out.owner));
+				// retrieve position of candidate in TopCandidates
+				if let Ok(i) = top_candidates.linear_search(&Stake {
+					owner: candidate.clone(),
+					amount: new_total,
+				}) {
+					// update TotalStake only if candidate becomes collator
+					let max_selected_candidates = MaxSelectedCandidates::<T>::get().saturated_into::<usize>();
+					if i < max_selected_candidates {
+						// candidate either pushed out the least staked collator or updated their own
+						// stake
```

Comment:
Yes, but the value is of type `Stake` and not `Balance` or `AccountId`. Thus, by increasing the staked amount, the candidate can replace their own entry in the TopCandidates list. Otherwise, the candidate can replace any other candidate with less stake.

Reply:
1cf58e8